src/binding.cc

```diff
@@ -26,88 +25,69 @@
+    exports.Set(Napi::String::New(env, "convert"),
+                Napi::Function::New(env, Convert));
 #define EXPORT_ERRNO(err) \
-    obj->Set(Nan::New<String>(#err).ToLocalChecked(), Nan::New<Integer>(err))
+    exports.Set(Napi::String::New(env, #err), Napi::Number::New(env, err))
     EXPORT_ERRNO(EINVAL);
     EXPORT_ERRNO(EILSEQ);
     EXPORT_ERRNO(E2BIG);
 #undef EXPORT_ERRNO
+    return exports;
   }
 
-  static NAN_METHOD(Make)
+  static Napi::Value Make(const Napi::CallbackInfo& info)
   {
-    Nan::Utf8String from_encoding(info[0]);
-    Nan::Utf8String to_encoding(info[1]);
-    iconv_t conv = iconv_open(*to_encoding, *from_encoding);
+    std::string from_encoding = info[0].As<Napi::String>();
+    std::string to_encoding = info[1].As<Napi::String>();
+    iconv_t conv = iconv_open(to_encoding.c_str(), from_encoding.c_str());
     if (conv == reinterpret_cast<iconv_t>(-1)) {
-      return info.GetReturnValue().SetNull();
+      return info.Env().Null();
     }
-    Iconv* iv = new Iconv(conv);
-    Local<Object> obj =
-        Nan::New<ObjectTemplate>(object_template)->NewInstance();
-    Nan::SetInternalFieldPointer(obj, 0, iv);
-    Nan::Persistent<Object> persistent(obj);
-    persistent.SetWeak(iv, WeakCallback, Nan::WeakCallbackType::kParameter);
-    info.GetReturnValue().Set(obj);
+    Napi::Value param = Napi::External<void>::New(info.Env(), conv);
+    return Iconv::object_template.New({param});
   }
 
-  static NAN_METHOD(Convert)
+  static Napi::Value Convert(const Napi::CallbackInfo& info)
   {
-    Iconv* iv = static_cast<Iconv*>(
-        Nan::GetInternalFieldPointer(info[0].As<Object>(), 0));
-    const bool is_flush = Nan::To<bool>(info[8]).FromJust();
+    Napi::Env env = info.Env();
+    Iconv* iv = Iconv::Unwrap(info[0].As<Napi::Object>());
+
+    const bool is_flush = info[8].As<Napi::Boolean>().Value();
     ICONV_CONST char* input_buf =
-        is_flush ? NULL : node::Buffer::Data(info[1].As<Object>());
-    size_t input_start = Nan::To<uint32_t>(info[2]).FromJust();
-    size_t input_size = Nan::To<uint32_t>(info[3]).FromJust();
-    char* output_buf = node::Buffer::Data(info[4].As<Object>());
-    size_t output_start = Nan::To<uint32_t>(info[5]).FromJust();
-    size_t output_size = Nan::To<uint32_t>(info[6]).FromJust();
-    Local<Array> rc = info[7].As<Array>();
+        is_flush ? NULL : info[1].As<Napi::Buffer<ICONV_CONST char>>().Data();
+    size_t input_start = info[2].IsNumber() ? info[2].As<Napi::Number>().Uint32Value() : static_cast<uint32_t>(0);
+    size_t input_size = info[3].IsNumber() ? info[3].As<Napi::Number>().Uint32Value() : static_cast<uint32_t>(0);
+    char* output_buf = info[4].As<Napi::Buffer<char>>().Data();
+    size_t output_start = info[5].IsNumber() ? info[5].As<Napi::Number>().Uint32Value() : static_cast<uint32_t>(0);
+    size_t output_size = info[6].IsNumber() ? info[6].As<Napi::Number>().Uint32Value() : static_cast<uint32_t>(0);
```

Comment:
Correct me if I'm wrong but I don't think you don't need the ternaries here? Coercing any value to uint32_t either returns the numeric value or zero on error.

If you're worried about failed coercion throwing an exception, then that's probably best handled by coercing in lib/iconv.js before passing it onto C++ land.

Reply:
The problem here is that if you don't pass a Number to the native side you will get an exception so I modified the JavaScript side like you suggested.